packages/flutter/lib/src/material/tooltip.dart

```diff
@@ -0,0 +1,295 @@
+      _entry = new OverlayEntry(builder: (BuildContext context) {
+        TextStyle textStyle = (config.style ?? Theme.of(context).text.body1).copyWith(color: config.textColor ?? Colors.white);
+        return new _TooltipOverlay(
+          message: config.message,
+          backgroundColor: config.backgroundColor ?? Colors.grey[700],
+          style: textStyle,
+          borderRadius: config.borderRadius,
+          height: config.height,
+          padding: config.padding,
+          opacity: config.opacity,
+          performance: _performance,
+          target: target,
+          verticalOffset: config.verticalOffset,
+          screenEdgeMargin: config.screenEdgeMargin,
+          preferBelow: config.preferBelow
+        );
+      });
+      Overlay.of(context).insert(_entry);
+    }
+    _timer?.cancel();
+    if (_performance.status != PerformanceStatus.completed) {
+      _timer = null;
+      _performance.forward();
+    } else {
+      resetShowTimer();
+    }
+  }
+
+  void hideTooltip() {
+    assert(_entry != null);
+    _timer?.cancel();
+    _timer = null;
+    _performance.reverse();    
+  }
+
+  void deactivate() {
+    if (_entry != null)
+      hideTooltip();
+    super.deactivate();
+  }
+
+  Widget build(BuildContext context) {
+    assert(Overlay.of(context) != null);
+    return new GestureDetector(
+      behavior: HitTestBehavior.opaque,
+      onLongPress: showTooltip,
+      child: config.child
+    );
+  }
+}
+
+class TooltipPositionDelegate extends OneChildLayoutDelegate {
+  TooltipPositionDelegate({
+    this.target,
+    this.verticalOffset,
+    this.screenEdgeMargin,
+    this.preferBelow
+  });
+  final Point target;
+  final double verticalOffset;
+  final EdgeDims screenEdgeMargin;
+  final bool preferBelow;
+
+  BoxConstraints getConstraintsForChild(BoxConstraints constraints) => constraints.loosen();
+
+  Offset getPositionForChild(Size size, Size childSize) {
+    // VERTICAL DIRECTION
+    bool tooltipBelow;
+    if (preferBelow) {
+      tooltipBelow = target.y + verticalOffset + childSize.height < size.height - screenEdgeMargin.bottom;
+    } else {
+      tooltipBelow = target.y - verticalOffset - childSize.height < screenEdgeMargin.top;
+    }
+    double y;
+    if (tooltipBelow)
+      y = math.min(target.y + verticalOffset, size.height - screenEdgeMargin.bottom);
+    else
+      y = math.max(target.y - verticalOffset - childSize.height, screenEdgeMargin.top);
+    // HORIZONTAL DIRECTION
+    double normalizedTargetX = target.x.clamp(screenEdgeMargin.left, size.width - screenEdgeMargin.right);
+    double x;
+    if (normalizedTargetX < screenEdgeMargin.left + childSize.width / 2.0) {
+      x = screenEdgeMargin.left;
+    } else if (normalizedTargetX > size.width - screenEdgeMargin.right - childSize.width / 2.0) {
+      x = size.width - screenEdgeMargin.right - childSize.width;
+    } else {
+      x = normalizedTargetX + childSize.width / 2.0;
+    }
+    return new Offset(x, y);
+  }
+
```

Comment:
I think that eventually we should have a test that verifies that tooltips whose components lie near the edges end up where they should.